go/felix/intdataplane/endpoint_mgr.go

```diff
@@ -173,70 +194,181 @@ func (m *endpointManager) OnUpdate(protoBufMsg interface{}) {
 			return
 		}
 		if msg.Addrs != nil {
-			m.ifaceAddrs[msg.Name] = msg.Addrs
+			m.hostIfaceToAddrs[msg.Name] = msg.Addrs
 		} else {
-			delete(m.ifaceAddrs, msg.Name)
+			delete(m.hostIfaceToAddrs, msg.Name)
 		}
-		m.dirtyHostEndpoints = true
+		m.hostEndpointsDirty = true
 	}
 }
 
 func (m *endpointManager) CompleteDeferredWork() error {
-	// Copy the pending interface updates to the active set.  Mark any interfaces that
-	// have come up to be reconfigured.
+	// Copy the pending interface state to the active set and mark any interfaces that have
+	// changed state for reconfiguration by resolveWorkload/HostEndpoints()
 	for ifaceName, state := range m.pendingIfaceUpdates {
 		if state == ifacemonitor.StateUp {
 			m.activeUpIfaces.Add(ifaceName)
-			m.ifaceNamesToReconfigure.Add(ifaceName)
+			if m.wlIfacesRegexp.MatchString(ifaceName) {
+				log.WithField("ifaceName", ifaceName).Info(
+					"Workload interface came up, marking for reconfiguration.")
+				m.wlIfaceNamesToReconfigure.Add(ifaceName)
+			}
 		} else {
 			m.activeUpIfaces.Discard(ifaceName)
 		}
-		logCxt := log.WithField("ifaceName", ifaceName)
-		if epID, ok := m.activeIfaceNameToID[ifaceName]; ok {
-			logCxt.Info("Interface state changed; marking for status update.")
-			m.ifaceIDsToUpdateStatus.Add(epID)
-		} else {
-			// We don't know about this interface yet (or it's already been deleted).
-			// If the endpoint gets created, we'll do the update then. If it's been
-			// deleted, we've already cleaned it up.
-			logCxt.Debug("Ignoring interface state change for unknown interface.")
-		}
+		// If this interface is linked to any already-existing endpoints, mark the endpoint
+		// status for recalculation.
+		m.markEndpointStatusDirtyByIface(ifaceName)
+	}
+
+	m.resolveWorkloadEndpoints()
+
+	if m.hostEndpointsDirty {
+		log.Debug("Host endpoints updated, resolving them.")
+		m.resolveHostEndpoints()
+		m.hostEndpointsDirty = false
+	}
+
+	// Since the pairings of endpoints and interfaces may have changed, rescan the interface
+	// state updates and mark endpoint statuses for update.
```

Comment:
I think you're right, it's over-defensive.  I don't think it can add anything new.  I'll remove it.